linfa-ensemble/src/random_forest/algorithm.rs

```diff
@@ -0,0 +1,191 @@
+        }
+        Ok(Array1::from(result))
+    }
+}
+
+impl ProbabilisticPredictor for RandomForest {
+    /// Return probability of predicted class for each sample, calculated as the rate of independent trees that
+    /// have agreed on such prediction
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A 2D array of floating point elements
+    ///
+    ///
+    fn predict_probabilities(
+        &self,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+    ) -> Result<Vec<Array1<f64>>, LinfaError> {
+        let ntrees = self.hyperparameters.n_estimators;
+        assert!(ntrees > 0, "Run .fit() method first");
+
+        let nclasses = self.trees[0].hyperparameters().n_classes;
+        let mut result: Vec<Array1<f64>> = Vec::with_capacity(x.nrows());
+
+        let flattened: Vec<Vec<u64>> = self
+            .trees
+            .iter()
+            .map(|tree| tree.predict(&x).unwrap().to_vec())
+            .collect();
+
+        for sample_idx in 0..x.nrows() {
+            let mut counter: Vec<u64> = vec![0; nclasses as usize];
+            for sp in &flattened {
+                // *counter_stats.entry(sp[sample_idx]).or_insert(0) += 1;
+                let single_pred = sp[sample_idx] as usize;
+                counter[single_pred] += 1;
+            }
+            let probas: Vec<f64> = counter.iter().map(|c| *c as f64 / ntrees as f64).collect();
+            result.push(Array1::from(probas));
+        }
+        Ok(result)
+    }
+}
+
+impl RandomForest {
+    pub fn fit(
+        hyperparameters: RandomForestParams,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+        y: &ArrayBase<impl Data<Elem = u64>, Ix1>,
+        max_n_rows: Option<usize>,
+    ) -> Self {
+        let n_estimators = hyperparameters.n_estimators;
+        let mut trees: Vec<DecisionTree> = Vec::with_capacity(n_estimators);
+        let single_tree_params = hyperparameters.tree_hyperparameters;
+        let max_n_rows = max_n_rows.unwrap_or_else(|| x.nrows());
+
+        //TODO check bootstrap
+        let _bootstrap = hyperparameters.bootstrap;
+        for _ in 0..n_estimators {
+            // Bagging here
+            let rnd_idx = Array::random((1, max_n_rows), Uniform::new(0, x.nrows())).into_raw_vec();
+            let xsample = x.select(Axis(0), &rnd_idx);
+            let ysample = y.select(Axis(0), &rnd_idx);
+            let tree = DecisionTree::fit(single_tree_params, &xsample, &ysample);
+            trees.push(tree);
+        }
+
+        Self {
+            hyperparameters,
+            trees,
+        }
+    }
+
+    /// Collect features from each tree in the forest and return hashmap(feature_idx: counts)
+    ///
+    pub fn feature_importances(&self) -> HashMap<usize, usize> {
+        let mut counter: HashMap<usize, usize> = HashMap::new();
+        for st in &self.trees {
+            // features in the single tree
+            let st_feats = st.features();
+            for f in st_feats.iter() {
```

Comment:
nit: As a user I would almost always want to see these ordered by index. Perhaps this should return a `Vec`?

Other note: there are alternative definitions of feature importance (e.g. Gini importance). I mention this not because we should implement that right away but because `usize` might not be the only sensible type. However that said, I would think that because these are pretty specific to random forest they can just be different methods with different return types.